streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/GraphGraceSearchUtil.java

```diff
@@ -78,7 +78,7 @@ private static Long extractGracePeriod(final StreamsGraphNode node) {
             } else if (processorSupplier instanceof KStreamSessionWindowAggregate) {
                 final KStreamSessionWindowAggregate kStreamSessionWindowAggregate = (KStreamSessionWindowAggregate) processorSupplier;
                 final SessionWindows windows = kStreamSessionWindowAggregate.windows();
-                return windows.gracePeriodMs();
+                return windows.gracePeriodMs() + windows.inactivityGap();
```

Comment:
Great catch! Was thinking if `max(grace, inactivity_gap)` is sufficient or not, but after a few more thoughts, I realize you are right :)

Reply:
Thanks for double-checking!

Reply:
> Otherwise, a record may arrive after the grace period, but before the inactivity gap, and merge with an existing window, moving the end time to the current time

This makes sense to me but thinking on this some more (and I may be missing something here), what does this imply for users that set a modest grace period but a larger inactivity gap? Wouldn't this allow records to come in beyond the users' expectations of the grace period?

But I'm guessing this is a better trade off than having a late arriving record adjust the end time of an existing window.

Reply:
I see Bill's concern here but I think it should be clarified: how exactly do we define the grace period (or how do we think users will interpret it?) -- is it a strict upper limit of allowable lateness, or a minimum amount of time we will wait? 

It seems tricky to determine the "correct" behavior when the grace period is < inactivity gap, since these two parameters essentially oppose each other Re: the inclusion of a record and so one has to trump the other when a record arrives between them.
For the more straightforward case grace period > inactivity gap though I'm not so sure about this logic ... it seems to me in this case we should (only) wait for the grace period.

Reply:
> is it a strict upper limit of allowable lateness, or a minimum amount of time we will wait?

It is a strict limit.

> Wouldn't this allow records to come in beyond the users' expectations of the grace period?

and

> It seems tricky to determine the "correct" behavior when the grace period is < inactivity gap, since these two parameters essentially oppose each other Re: the inclusion of a record and so one has to trump the other when a record arrives between them.

I don't think so. The grace period is additive to the session-gap. If grace period is zero, and gap=5000, and we have a single-record session at timestamp 1000, we would emit it at 6000 (=end+gap+grace)

> For the more straightforward case grace period > inactivity gap though I'm not so sure about this logic ... it seems to me in this case we should (only) wait for the grace period

From my understanding, for this case we would emit after gap plus grace period passed.

Note, that as long as the gap time did not elapse, _ordered_ records can expand the session. In our example above, we had a single-record session at 1000 (let's say key=A) with gap 5000. If a record with key=A arrives at 3000, the session is expended to `<A,[1000,3000]>` and the new "gap cut-off point" moves from 6000 to 8000. We need to allow for this case even if grace period is zero: the record was not out-of-order! However, if we get a record with different key=B at timestamp 9000, and afterwards a record with key=A and timestamps 7000 arrives, we don't add it to the old session of `[1000,3000]` because the new record was out-of-order and missed the grace period (and it dropped as late record). In order to incorporate the record, we would need a grace period of 1000: for this case, key=B (ts: 9000) is just one tick before grace period passed for the open A-session, and we don't close the A-session yet, and afterward extend the A-session to `[1000,7000]` to incorporate the out-of-order record, resulting in new session-gap cut-off time of 12000 and new close time of 13000 (incorporating the 1000ms grace period).

Reply:
Thanks, all. I agree it's a bit mind-bending (hence, the bug).

@mjsax is right, it's not a "loose" definition at all. It's both strict and well defined. There's bound to be some confusion, because Streams differs from other SP frameworks in the way we handle late records. (In my opinion, our way is better ;) ) Building on Matthias's example, I'll try and give a high-level "mental model" of what's going on here.

Some quick definitions, since all the SP frameworks and papers like to use overlapping terms differently:
* "stream time" is "right now" with respect to the data you've seen so far. Notably, Streams defines "stream time" as the maximum observed timestamp so far in the stream (specifically, in a topic-partition). Beam (for example) defines it using a "watermark", which dynamically lags the maximum observed timestamp, in an effort to not have late events at all (but you have future events instead, which I think is kind of weird).
* "lateness" (specifically for the purpose of this conversation) is "stream time - record timestamp". Note that in a perfectly-ordered stream, all records have the maximum observed timestamp so far, so they all have zero lateness. In Streams, any record that has an out-of-order timestamp has some amount of lateness. Also, note that in Beam (for example), records may have out-of-order timestamps, but not be considered "late" at all because the watermark lags the max observed timestamp.

That was fun.

So, in Beam (and company), you can "bound the maximum allowed lateness" explicitly, and the default bound is "zero". But you still have out-of-order events, because of the fancy stream-time/watermark definition. But in Streams, you cannot! And you don't need to! I don't think this is a useful property to define for an SP application. Why bother bounding the maximum allowed lateness of your application? What can you do with this information?

What matters, I think, is stuff like bounding the time you have to wait for final updates, or the amount of time you're going to keep the window state stored (either for updates or for query). Note that these bounds are specifically on _windows_, not _records_. That's why our "grace period" is set on the window definition, and the "retention time" is set on the window store, and there is no mention of a record-lateness bound anywhere.

So the right way to think about the grace period is bounding the time (in stream time) that the window stays open for updates, and hence the amount of time you have to wait for the "final updates", if you're using suppression. This doesn't put any specific bound on record lateness, since a record that "should" be toward the beginning of the window is admitted with a higher lateness than a record that "should" be toward the end of the window.

But! It's also not quite perfect to say that it's the "amount of time we'll wait", since that phrase leaves out the time you start waiting... Is it the beginning of the window, the end of the window, or what? If you imagine a perfectly ordered stream, then there is some point at which you know a window won't be updated anymore. We can call this the *natural close time*. For fixed-size windows, this is the window end time. For data-driven windows, it's more nebulous, because it depends on the logic of the window definition. Specifically, for session windows, it's the end-time plus the inactivity gap.

For _im_perfectly ordered streams (ones with out-of-order/late records), we want to keep the window open for updates for a little longer (the grace period), so the actual *close time* of the window is the natural close time + the grace period. In other words, I'd modify @ableegoldman 's phrase and say that the grace priod is the "amount of time we'll wait for updates after the natural close time of the window".

So, in response to Bill, yes this behavior might violate users' expectations about what happens, but only if they don't understand what grace period means. It could be that less terse documentation is in order.

Reply:
With the above context in mind, I think I see why I found the code here a little sketchy. The associativity is messed up. It should be `(window end + inactivity gap) + grace`. What I mean is that `(window end + inactivity gap)` is an identifiable concept (the natural close time).

But the code is actually `window end + (inactivity gap + grace)`, which feels wrong because `(inactivity gap + grace)` is just two numbers added together. It's algebraically correct, but it's not semantically aligned.

But the structure of the code makes the first association more awkward to implement, since both the inactivity gap and the grace period are static properties of the window definition, while the window end time is unique to each window... I'll see if I can re-structure the code to be more intuitive, but it might be that we just have to live with it.

Reply:
Thanks @vvcephei for the detailed analysis! In light of a more concrete understanding of grace period and window end/close time the situation is much less ambiguous...maybe you should be writing all our docs :wink:

Reply:
@ableegoldman You must be aware that @vvcephei holds a degree for that :P